{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c29bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install dependencies for phylogenetic analysis \n",
    "conda create -n phylo_te -c bioconda -c conda-forge roary iqtree mafft fasttree prokka\n",
    "\n",
    "conda activate phylo_te\n",
    "\n",
    "# Verify installation of dependencies \n",
    "roary -v\n",
    "iqtree -v\n",
    "mafft --version\n",
    "fasttree -help\n",
    "prokka --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "199c19d2",
   "metadata": {},
   "source": [
    "### Roary para analise de pangenoma e geracao de alinhamento.\n",
    "\n",
    "### IQ-TREE para construcao da arvore filogenetica.\n",
    "\n",
    "### iTOL para visualizacao da arvore e anotacao das IS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bef5761d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. anotar genes\n",
    "prokka --outdir GCF_001_prokka --prefix GCF_001 GCF_001.fasta\n",
    "\n",
    "## --outdir eh o diretorio de saida, ou seja, onde serao armazenados os arquivos de saida\n",
    "## --prefix eh o prefixo dos arquivos de saida, ou seja, o nome que sera dado aos arquivos\n",
    " \n",
    "## NCBI.fasta eh o arquivo de entrada, ou seja, o arquivo que sera anotado, que e o arquivo do genoma do NCBI\n",
    "\n",
    "### Fazer esta etapa para cada genoma que for ser analisado, ou seja, para cada genoma do NCBI\n",
    "\n",
    "# MODO AUTOMATICO #\n",
    "## Abra o nano ##\n",
    "for file in *.fasta; do\n",
    "    base=$(basename \"$file\" .fasta)\n",
    "    prokka --outdir \"${base}_prokka\" --prefix \"$base\" \"$file\"\n",
    "done\n",
    "### copiar e colar o comando acima \n",
    "#### Aperte Ctrl + o para salvar, salvar como run_prokka.sh, aperte Enter para salvar\n",
    "##### Aperte Ctrl + x para sair do nano\n",
    "###### Agora e so rodar o script ---> bash run_prokka.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b36886e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# criar um diretorio para a coleção de arquivos .gff gerados\n",
    "cd /mnt/dados/victor_baca/outputs/Phylogenetic_analysis/iq-tree\n",
    "mkdir -p gff_collection\n",
    "\n",
    "# copiar os arquivos .gff gerados no diretorio gff_collection\n",
    "for diretorio in *_prokka; do\n",
    "    cp \"$diretorio\"/*.gff /mnt/dados/victor_baca/outputs/Phylogenetic_analysis/iq-tree/gff_collection/\n",
    "done \n",
    "\n",
    "## 2. rodar o roary\n",
    "roary -e -n -v -f roary_output *.gff\n",
    "### -e: usa alinhamento de nucleotideos \n",
    "### -n: nao filtra para apenas um representante por gene \n",
    "### -v: modo verboso \n",
    "### -f: pasta de saida\n",
    "## coloca o gff do prokka fez\n",
    "### vai gerar um arquivo .csv e um .aln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8957675",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 3. rodar o iq-tree\n",
    "cd /mnt/dados/victor_baca/outputs/Phylogenetic_analysis/iq-tree/gff_collection/roary_output\n",
    "iqtree -s core_gene_alignment.aln -nt AUTO -m GTR+G -bb 1000\n",
    "###vai criar um arquivo core_gene_alignment.aln.treefile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "318b34bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. criar um novo arquivo .tsv com os aarquivos .tsv do ISEScan\n",
    "''' genome  IS3     IS256   IS30\n",
    "    GCF_001 2       0       1\n",
    "    GCF_002 1       1       0\n",
    "    GCF_003 3       2       1'''\n",
    "\n",
    "# SCRIPT PARA GERAR O ARQUIVO .TSV COM AS IS E AS CEPAS ###########################################################\n",
    "## Para utilizar este script, é necessário ter os arquivos .tsv do ISEScan numa pasta unica.######################\n",
    "### Ou seja, todos os arquivos .tsv do ISEScan devem estar na mesma pasta.########################################\n",
    "#### Codigo para fazer a copia de todos os arquivos .tsv de um determinado diretorio para outro (sem importar se esta em subdiretorios)\n",
    "find '/mnt/dados/victor_baca/outputs/isescan' -type f -name \"*.tsv\" -exec cp -t '/mnt/dados/victor_baca/outputs/Phylogenetic_analysis/iq-tree/tsv' {} +\n",
    "##### Para rodar o script, pois a variavel \"pasta_tsv\" deve ser alterada para o caminho da pasta onde estão os arquivos .tsv do ISEScan.##########\n",
    "import csv\n",
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "# Pasta onde estão os arquivos .tsv do ISEScan\n",
    "pasta_tsv = \"/mnt/dados/victor_baca/outputs/Phylogenetic_analysis/iq-tree/tsv\"\n",
    "\n",
    "# Inicializa estrutura para armazenar dados\n",
    "dados_gerais = defaultdict(lambda: defaultdict(int))\n",
    "familias_todas = set()\n",
    "\n",
    "# Percorre todos os arquivos .tsv\n",
    "for arquivo in os.listdir(pasta_tsv):\n",
    "    if not arquivo.endswith(\".tsv\"):\n",
    "        continue\n",
    "\n",
    "    caminho = os.path.join(pasta_tsv, arquivo)\n",
    "    cepa = os.path.splitext(arquivo)[0]  # ex: strepto1_isescan → strepto1_isescan\n",
    "\n",
    "    # Lê o arquivo\n",
    "    with open(caminho) as f:\n",
    "        reader = csv.DictReader(f, delimiter='\\t')\n",
    "        for linha in reader:\n",
    "            familia = linha['family']\n",
    "            dados_gerais[cepa][familia] += 1\n",
    "            familias_todas.add(familia)\n",
    "\n",
    "# Ordena as famílias\n",
    "familias_ordenadas = sorted(familias_todas)\n",
    "\n",
    "# Escreve arquivo de saída\n",
    "with open(\"is_counts.tsv\", \"w\", newline='') as f_out:\n",
    "    writer = csv.writer(f_out, delimiter='\\t')\n",
    "    writer.writerow([\"genome\"] + familias_ordenadas)\n",
    "\n",
    "    for cepa in sorted(dados_gerais):\n",
    "        linha = [cepa]\n",
    "        for fam in familias_ordenadas:\n",
    "            linha.append(dados_gerais[cepa].get(fam, 0))\n",
    "        writer.writerow(linha)\n",
    "\n",
    "print(\"Arquivo 'is_counts.tsv' gerado com sucesso!\")\n",
    "\n",
    "############################################################## output -> is_counts.tsv\n",
    "\n",
    "# Apos isso transforme o arquivo tsv gerado pelo script em um arquivo txt, com o script abaixo\n",
    "################################################################################################\n",
    "import csv\n",
    "\n",
    "input_file = \"is_counts.tsv\"\n",
    "output_file = \"itol_is_binary.txt\"\n",
    "\n",
    "# Parâmetros visuais\n",
    "shape = \"2\"  # quadrado\n",
    "cores_padrao = [\"#e6194b\", \"#3cb44b\", \"#ffe119\", \"#4363d8\", \"#f58231\", \"#911eb4\", \"#46f0f0\", \"#f032e6\", \"#bcf60c\", \"#fabebe\"]\n",
    "\n",
    "with open(input_file) as f:\n",
    "    reader = csv.reader(f, delimiter='\\t')\n",
    "    header = next(reader)\n",
    "    familias = header[1:]  # Ignora 'genome'\n",
    "    \n",
    "    # Gera cores (repete se faltar)\n",
    "    cores = (cores_padrao * ((len(familias) // len(cores_padrao)) + 1))[:len(familias)]\n",
    "    formas = [shape] * len(familias)\n",
    "    \n",
    "    dados = list(reader)\n",
    "\n",
    "with open(output_file, 'w', newline='') as out:\n",
    "    out.write(\"DATASET_BINARY\\n\")\n",
    "    out.write(\"SEPARATOR TAB\\n\")\n",
    "    out.write(\"DATASET_LABEL\\tIS elements\\n\")\n",
    "    out.write(\"FIELD_LABELS\\t\" + \"\\t\".join(familias) + \"\\n\")\n",
    "    out.write(\"FIELD_COLORS\\t\" + \"\\t\".join(cores) + \"\\n\")\n",
    "    out.write(\"FIELD_SHAPES\\t\" + \"\\t\".join(formas) + \"\\n\")\n",
    "    out.write(\"DATA\\n\")\n",
    "\n",
    "    for linha in dados:\n",
    "        out.write(\"\\t\".join(linha) + \"\\n\")\n",
    "\n",
    "################################################################################### output -> itol_is_binary.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37d8f2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### 5. Visualizar no iTOL\n",
    "## Va para https://itol.embl.de\n",
    "## Clique em “Upload Tree”\n",
    "## Suba o arquivo: core_gene_alignment.aln.treefile\n",
    "## Clique em “Datasets” > “Upload” > selecione o .txt com ISs do arquivo .tsv (itol_is_binary.txt)\n",
    "## Visualize com circulos coloridos ou heatmap por IS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03132edb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copiar um arquivo do servidor para maquina local (windowns)\n",
    "scp victorbaca@10.20.34.31:/mnt/dados/victor_baca/outputs/Phylogenetic_analysis/iq-tree/tsv/itol_is_binary.txt \"C:\\Users\\vitor\\Downloads\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "novoenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
